docs/modules/candid-guide/index.adoc

```diff
@@ -0,0 +1,1098 @@
+
+`Int`
+
+Corresponding Rust type::
+
+TODO
+
+Corresponding JavaScript values::
+
+TODO
+
+
+[#type-natN]
+[#type-intN]
+=== Type natN and intN
+
+The types `nat8`, `nat16`, `nat32`, `nat64`, `int8`, `int16`, `int32` and `nat64` represent numbers with a representation of that many bits, and can be used in more “low-level” interface.
+
+The range of `natN` is \(\{0 \ldots 2^{N}-1\}\), and the range of `intN` is \(\{-2^{N-1} \ldots 2^{N-1}-1\}\).
+
+The on-wire representation is exactly that many bits long. So for small values, `nat` is more space-efficent than `nat64`.
+
+Type syntax::
+
+`nat8`, `nat16`, `nat32`, `nat64`, `int8`, `int16`, `int32` or `nat64`
+
+Textual syntax::
+
+Same as `nat` (for `natN`) resp. `int` (for `intN`).
+
+Corresponding Motoko type::
+
+`natN` translate by default to `NatN`, but can also correspond got `WordN`
++
+`intN` translate to `IntN`.
+
+Corresponding Rust type::
+
+TODO
+
+Corresponding JavaScript values::
+
+TODO
+
+[#type-floatN]
+=== Type float16 and float32
+
+The types `float16`, `float32` represent IEEE 754 floating point numbers, in single precision (32 bit) and double precision (64 bit).
+
+Type syntax::
+
+`float16`, `float32`
+
+Textual syntax::
+
+The same syntax as `int`, plus floating point literals as follows
++
+....
+1245.678
++1245.678
+-1_000_000.000_001
+34e10
+34E+10
+34e-10
+0xDEAD.BEEF
+0xDEAD.BEEFP-10
+0xDEAD.BEEFp+10
+....
+
+Corresponding Motoko type::
+
+`float64` corresponds to `Float`.
++
+`float32` does _not_ have a representation in Motoko. Candid interfaces using `float32` cannot be served from or used from Motoko programs.
+
+Corresponding Rust type::
+
+TODO
+
+Corresponding JavaScript values::
+
+TODO
+
+[#type-bool]
+=== Type bool
+
+The type of truths values.
+
+Type syntax::
+
+`bool`
+
+Textual syntax::
+
+`true`, `false`
+
+Corresponding Motoko type::
+
+`Bool`
+
+Corresponding Rust type::
+
+TODO
+
+Corresponding JavaScript values::
+
+TODO
+
+[#type-vec]
+=== Type vec t
+
+The `vec` type representes vectors (sequences, lists, arrays): A value of type `vec t` contains zero or more values of type `t` in a specific order
+
+Type syntax::
+
+`vec bool`, `vec nat8`, `vec vec text`,…
+
+Textual syntax::
++
+....
+vec {}
+vec { "john@doe.com"; "john.doe@example.com" };
+....
+
+Subtypes::
+
+* Whenever `t` is a subtype of `t'`, then `vec t` is a subtype of `vec t'`.
+* `blob` is a subtype of `vec nat8`
+
+Supertypes::
+
+* Whenever `t` is a supertype of `t'`, then `vec t` is a supertype of `vec t'`.
+* `blob` is a supertype of `vec nat8`
+
+Corresponding Motoko type::
+
+`[T]`, where the Motoko type `T` corresponds to `t`
+
+Corresponding Rust type::
+
+`Vec<T>`, where the Rust type `T` corresponds to `t`
+
+Corresponding JavaScript values::
+
+TODO:
+
+[#type-opt]
+=== Type opt t
+
+The `opt t` type contains of the values of type `t`, plus the special `null` value. It is used to express that some value is optional, i.e. it may be present, or there may be no value.
```

Comment:
```suggestion
The `opt t` type contains all the values of type `t`, plus the special `null` value. It is used to express that some value is optional, i.e. it may be present as some value of type `t`, or absent as the value `null`.
```